File: frontend/lib/skins/namePreview.ts
import type { CSSProperties } from "react";

/** Approximate TLibs / Minecraft name colour preview for the upload form. */

export const NAME_STYLES = [
  "bold",
  "italic",
  "underline",
  "strikethrough",
] as const;

export type NameStyle = (typeof NAME_STYLES)[number];

export const LEGACY_PALETTE: { code: string; hex: string; label: string }[] = [
  { code: "0", hex: "#000000", label: "Black" },
  { code: "1", hex: "#0000aa", label: "Dark blue" },
  { code: "2", hex: "#00aa00", label: "Dark green" },
  { code: "3", hex: "#00aaaa", label: "Dark aqua" },
  { code: "4", hex: "#aa0000", label: "Dark red" },
  { code: "5", hex: "#aa00aa", label: "Dark purple" },
  { code: "6", hex: "#ffaa00", label: "Gold" },
  { code: "7", hex: "#aaaaaa", label: "Gray" },
  { code: "8", hex: "#555555", label: "Dark gray" },
  { code: "9", hex: "#5555ff", label: "Blue" },
  { code: "a", hex: "#55ff55", label: "Green" },
  { code: "b", hex: "#55ffff", label: "Aqua" },
  { code: "c", hex: "#ff5555", label: "Red" },
  { code: "d", hex: "#ff55ff", label: "Light purple" },
  { code: "e", hex: "#ffff55", label: "Yellow" },
  { code: "f", hex: "#ffffff", label: "White" },
];

const HEX_RE = /^#?[0-9A-Fa-f]{6}$/;

export function normalizePreviewHex(token: string): string | null {
  const t = token.trim();
  if (!t) return null;
  if (t.length === 2 && (t[0] === "§" || t[0] === "&")) {
    const hit = LEGACY_PALETTE.find((p) => p.code === t[1].toLowerCase());
    return hit?.hex ?? null;
  }
  if (t.length === 1) {
    const hit = LEGACY_PALETTE.find((p) => p.code === t.toLowerCase());
    return hit?.hex ?? null;
  }
  if (!HEX_RE.test(t)) return null;
  return (t.startsWith("#") ? t : `#${t}`).toLowerCase();
}

function parseRgb(hex: string): [number, number, number] {
  const v = parseInt(hex.slice(1), 16);
  return [(v >> 16) & 0xff, (v >> 8) & 0xff, v & 0xff];
}

function rgbToHex(r: number, g: number, b: number): string {
  return (
    "#" +
    [r, g, b]
      .map((n) => Math.round(n).toString(16).padStart(2, "0"))
      .join("")
  );
}

function lerp(
  from: [number, number, number],
  to: [number, number, number],
  t: number
): string {
  const c = Math.max(0, Math.min(1, t));
  return rgbToHex(
    from[0] + (to[0] - from[0]) * c,
    from[1] + (to[1] - from[1]) * c,
    from[2] + (to[2] - from[2]) * c
  );
}

export type PreviewSpan = { char: string; color: string };

/** Per-character colours matching TLibs applyColourGradient. */
export function previewSpans(
  text: string,
  colourTokens: string[]
): PreviewSpan[] {
  const plain = text || "";
  const hexes = colourTokens
    .map(normalizePreviewHex)
    .filter((h): h is string => Boolean(h));
  if (!plain) return [];
  if (hexes.length === 0) {
    return [...plain].map((char) => ({ char, color: "#ffffff" }));
  }
  if (hexes.length === 1) {
    return [...plain].map((char) => ({ char, color: hexes[0] }));
  }
  const stops = hexes.length;
  const length = plain.length;
  return [...plain].map((char, i) => {
    const t = length === 1 ? 0 : i / (length - 1);
    let segment = Math.floor(t * (stops - 1));
    if (segment >= stops - 1) segment = stops - 2;
    const localT = t * (stops - 1) - segment;
    return {
      char,
      color: lerp(
        parseRgb(hexes[segment]),
        parseRgb(hexes[segment + 1]),
        localT
      ),
    };
  });
}

export function previewStyleCss(styles: string[]): CSSProperties {
  const set = new Set(styles.map((s) => s.toLowerCase()));
  const deco = [
    set.has("underline") || set.has("underlined") ? "underline" : "",
    set.has("strikethrough") || set.has("strike") ? "line-through" : "",
  ]
    .filter(Boolean)
    .join(" ");
  return {
    fontWeight: set.has("bold") ? 700 : 400,
    fontStyle: set.has("italic") ? "italic" : "normal",
    textDecoration: deco || "none",
  };
}

File: backend/src/skins/name_preview.py
"""Render coloured display-name preview PNGs for Discord staff review."""

from __future__ import annotations

import io
import re
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont

from .db import SKINS_DIR

PAD_X = 16
PAD_Y = 12
BG = (26, 26, 26)
DEFAULT_HEX = "#ffffff"
FONT_SIZE = 28
_HEX_RE = re.compile(r"^#?[0-9A-Fa-f]{6}$")

LEGACY_HEX = {
    "0": "#000000",
    "1": "#0000aa",
    "2": "#00aa00",
    "3": "#00aaaa",
    "4": "#aa0000",
    "5": "#aa00aa",
    "6": "#ffaa00",
    "7": "#aaaaaa",
    "8": "#555555",
    "9": "#5555ff",
    "a": "#55ff55",
    "b": "#55ffff",
    "c": "#ff5555",
    "d": "#ff55ff",
    "e": "#ffff55",
    "f": "#ffffff",
}


def _font(size: int = FONT_SIZE) -> ImageFont.ImageFont:
    try:
        return ImageFont.truetype("arial.ttf", size)
    except OSError:
        try:
            return ImageFont.truetype("DejaVuSans.ttf", size)
        except OSError:
            return ImageFont.load_default()


def _normalize_hex(token: str) -> str | None:
    t = (token or "").strip()
    if not t:
        return None
    if len(t) == 2 and t[0] in ("§", "&", "\u00a7"):
        return LEGACY_HEX.get(t[1].lower())
    if len(t) == 1:
        return LEGACY_HEX.get(t.lower())
    if t.startswith("\u00a7") and len(t) == 2:
        return LEGACY_HEX.get(t[1].lower())
    if not _HEX_RE.match(t):
        return None
    return (t if t.startswith("#") else f"#{t}").lower()


def _parse_rgb(hex_color: str) -> tuple[int, int, int]:
    v = int(hex_color[1:], 16)
    return (v >> 16) & 0xFF, (v >> 8) & 0xFF, v & 0xFF


def _lerp(
    a: tuple[int, int, int], b: tuple[int, int, int], t: float
) -> tuple[int, int, int]:
    c = max(0.0, min(1.0, t))
    return (
        int(round(a[0] + (b[0] - a[0]) * c)),
        int(round(a[1] + (b[1] - a[1]) * c)),
        int(round(a[2] + (b[2] - a[2]) * c)),
    )


def _char_colours(text: str, colour_tokens: list[str] | None) -> list[tuple[str, tuple[int, int, int]]]:
    plain = text or ""
    hexes: list[str] = []
    for tok in colour_tokens or []:
        h = _normalize_hex(str(tok))
        if h:
            hexes.append(h)
    if not plain:
        return []
    if not hexes:
        rgb = _parse_rgb(DEFAULT_HEX)
        return [(ch, rgb) for ch in plain]
    if len(hexes) == 1:
        rgb = _parse_rgb(hexes[0])
        return [(ch, rgb) for ch in plain]
    stops = len(hexes)
    length = len(plain)
    out: list[tuple[str, tuple[int, int, int]]] = []
    for i, ch in enumerate(plain):
        t = 0.0 if length == 1 else i / (length - 1)
        segment = int(t * (stops - 1))
        if segment >= stops - 1:
            segment = stops - 2
        local_t = t * (stops - 1) - segment
        out.append(
            (
                ch,
                _lerp(
                    _parse_rgb(hexes[segment]),
                    _parse_rgb(hexes[segment + 1]),
                    local_t,
                ),
            )
        )
    return out


def render_name_preview_png(
    display_name: str,
    name_colours: list[str] | None = None,
    name_styles: list[str] | None = None,
) -> bytes:
    """Return PNG bytes of the coloured display name on a dark background."""
    text = (display_name or "").strip() or "—"
    styles = {str(s).strip().lower() for s in (name_styles or [])}
    bold = "bold" in styles
    italic = "italic" in styles
    underline = "underline" in styles or "underlined" in styles
    strike = "strikethrough" in styles or "strike" in styles

    font = _font(FONT_SIZE + (2 if bold else 0))
    chars = _char_colours(text, name_colours)

    # Measure with a temp image
    tmp = Image.new("RGB", (1, 1), BG)
    draw = ImageDraw.Draw(tmp)
    widths: list[int] = []
    max_h = 0
    for ch, _rgb in chars:
        bbox = draw.textbbox((0, 0), ch, font=font)
        w = max(1, bbox[2] - bbox[0])
        h = max(1, bbox[3] - bbox[1])
        widths.append(w)
        max_h = max(max_h, h)

    text_w = sum(widths) + max(0, len(widths) - 1)  # 1px letter spacing
    # Italic shear needs a little extra room
    if italic:
        text_w = int(text_w * 1.08) + 4

    width = text_w + PAD_X * 2
    height = max_h + PAD_Y * 2 + (4 if underline or strike else 0)
    canvas = Image.new("RGB", (max(width, 64), max(height, 40)), BG)
    draw = ImageDraw.Draw(canvas)

    x = PAD_X
    baseline_y = PAD_Y
    for (ch, rgb), w in zip(chars, widths):
        if italic:
            # Draw onto a small tile then shear
            tile = Image.new("RGBA", (w + 8, max_h + 8), (0, 0, 0, 0))
            td = ImageDraw.Draw(tile)
            td.text((2, 2), ch, fill=rgb + (255,), font=font)
            shear = 0.25
            sheared = tile.transform(
                tile.size,
                Image.Transform.AFFINE,
                (1, shear, -shear * 4, 0, 1, 0),
                resample=Image.Resampling.BILINEAR,
            )
            canvas.paste(sheared, (x - 2, baseline_y - 2), sheared)
        else:
            if bold:
                for dx, dy in ((0, 0), (1, 0), (0, 1)):
                    draw.text((x + dx, baseline_y + dy), ch, fill=rgb, font=font)
            else:
                draw.text((x, baseline_y), ch, fill=rgb, font=font)
        x += w + 1

    line_y_base = baseline_y + max_h
    if underline:
        draw.line(
            (PAD_X, line_y_base + 1, PAD_X + text_w, line_y_base + 1),
            fill=(220, 220, 220),
            width=2,
        )
    if strike:
        mid = baseline_y + max_h // 2
        draw.line(
            (PAD_X, mid, PAD_X + text_w, mid),
            fill=(220, 220, 220),
            width=2,
        )

    buf = io.BytesIO()
    canvas.save(buf, format="PNG")
    return buf.getvalue()


def write_name_preview(
    submission_id: str,
    display_name: str,
    name_colours: list[str] | None = None,
    name_styles: list[str] | None = None,
) -> Path:
    out_dir = SKINS_DIR / submission_id
    out_dir.mkdir(parents=True, exist_ok=True)
    path = out_dir / "name_preview.png"
    path.write_bytes(
        render_name_preview_png(display_name, name_colours, name_styles)
    )
    return path
